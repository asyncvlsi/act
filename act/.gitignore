html
latex
libact_*.a
i386_*
*.o
act_parse*
act.cy
act_walk_X*
expr.c
expr.h
act_parse_id.h
Makefile.deps
NOTES
TAGS
*~
Changelog
