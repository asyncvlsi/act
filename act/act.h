/*************************************************************************
 *
 *  Copyright (c) 2011-2018 Rajit Manohar
 *  All Rights Reserved
 *
 **************************************************************************
 */
#ifndef __ACT_H__
#define __ACT_H__

#include <act/types.h>
#include <act/lang.h>

/**
 * @file act.h
 *       Contains top-level initialization/management functions for
 *       the ACT library
 */

class Act {
 public:
  /**
   * Initialize the ACT library
   *
   *  @param argc is a pointer to argc (command-line processing)
   *  @param argv is a pointer to argv (command-line processing)
   *
   *  @return *argc and *argv are modified to reflect the command-line
   *  options left after ACT has extracted the ones it understands
   */
  static void Init (int *argc, char ***argv);


  /**
   * Maximum depth of recursion for expanding types
   */
  static int max_recurse_depth;

  /**
   * Create an act data structure for the specified input file
   *
   * @param s is the name of the file containing the top-level ACT
   */
  Act (const char *s = NULL);
  ~Act ();


  /**
   * Expand types!
   */
  void Expand ();

  
  void mangle (char *s);	// install string mangling functions
  int mangle_active() { return any_mangling; }
  int mangle_string (char *src, char *dst, int sz);
  int unmangle_string (char *src, char *dst, int sz);
  void mfprintf (FILE *fp, const char *s, ...);
  void ufprintf (FILE *fp, const char *s, ...);
  int msnprintf (char *fp, int sz, const char *s, ...);
  int usnprintf (char *fp, int sz, const char *s, ...);


  /* 
     API functions
  */
  Process *findProcess (const char *s);
  Process *findProcess (ActNamespace *, const char *);
  ActNamespace *findNamespace (const char *s);
  ActNamespace *findNamespace (ActNamespace *, const char *);
  ActNamespace *Global() { return gns; }

  void *aux_find (const char *phase);
  void aux_add (const char *phase, void *data);

private:
  TypeFactory *tf;		/* type factory for the file */
  ActNamespace *gns;		/* global namespace */

  char mangle_characters[256];
  int inv_map[256];
  int any_mangling;

  struct Hashtable *aux;	// any aux storage you want
  
};

#endif /* __ACT_H__ */
