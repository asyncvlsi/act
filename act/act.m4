/*************************************************************************
 *
 *  Copyright (c) 2011-2018 Rajit Manohar
 *  All Rights Reserved
 *
 **************************************************************************
 */

/*------------------------------------------------------------------------
 *
 *  Act top-level pgen file
 *
 *------------------------------------------------------------------------
 */

%type[X] {{ ActTree ActRet }};

/*------------------------------------------------------------------------
 *
 * Grammar
 *
 *------------------------------------------------------------------------
 */
toplevel: [ imports_opens ] [ body ]
{{X:
    OPT_FREE ($1);

    if (!OPT_EMPTY ($2)) {
      ActRet *r;

      r = OPT_VALUE ($2);
      $A(r->type == R_ACT_BODY);
      $0->curns->setBody (r->u.body);
      FREE (r);
    }
    OPT_FREE ($2);
    return NULL;
}}
;

imports_opens
: import_open_item imports_opens
| import_open_item
;

body[ActBody *]
: body_item body
{{X:
    if ($1 == NULL) {
      return $2;
    }
    if ($2 == NULL) {
      return $1;
    }
    $1->Append ($2);
    return $1;
}}
| body_item
{{X: return $1; }}
;

ns_body[ActBody *]
: ns_body_item ns_body
{{X:
    if ($1 == NULL) {
      return $2;
    }
    if ($2 == NULL) {
      return $1;
    }
    $1->Append ($2);
    return $1;
}}
| ns_body_item
{{X:
    return $1;
}}
;

ns_body_item[ActBody *]
: definition
{{X:
    return NULL;
}}
| namespace_management
{{X:
    return NULL;
}}
| instance
{{X:
    ActBody_Inst *b;
    if ($0->override) {
      $E("Can't have override types at the top-level in a namespace");
    }
    $A($1);
    b = dynamic_cast <ActBody_Inst*>($1);
    $A(b);
    if (TypeFactory::isProcessType (b->BaseType())) {
      $E("Cannot instantiate processes within a namespace");
    }
    return $1;
}}
| connection
{{X:
    return $1;
}}
| alias
{{X:
    return $1;
}}
;

body_item[ActBody *]
: namespace_management
{{X: return NULL; }}
| base_item
{{X: return $1; }}
| definition
{{X: return NULL; }}
;

import_open_item: import_item  | open_item ;

/* namespaces */
include(namespaces.m4)

/* types */
include(types.m4)

/* definitions */
include(defs.m4)

/* languages */
include(lang.m4)

/* expressions */
include(expr.m4)
