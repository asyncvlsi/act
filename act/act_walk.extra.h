/*************************************************************************
 *
 *  Copyright (c) 2011 Rajit Manohar
 *  All Rights Reserved
 *
 **************************************************************************
 */
#ifndef __ACT_EXTRA_H__
#define __ACT_EXTRA_H__

#include <stdio.h>
#include <string.h>
#include "list.h"
#include "except.h"
#include "qops.h"
#include "misc.h"
#include <act/lang.h>
#include <act/path.h>
#include <act/namespaces.h>
#include <act/treetypes.h>
#include <act/types.h>
#include <act/inst.h>

#define OPT_EXISTS(x)    !list_isempty (x)
#define OPT_EMPTY(x)     list_isempty (x)
#define LIST_VALUE(x)     ((ActRet *)list_value (x))
#define OPT_VALUE(x)     LIST_VALUE (list_first (x))
#define OPT_FREE(x)      list_free(x)
#define OPT_VALUE2(x)    LIST_VALUE (list_next (list_first (x)))

Expr *const_expr (int val);
Expr *act_walk_X_expr (ActTree *cookie, Expr *e);
void print_ns_string (FILE *fp, list_t *l);

#endif /* __ACT_EXTRA_H__ */
