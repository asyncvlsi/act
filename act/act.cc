/*************************************************************************
 *
 *  Copyright (c) 2017-2018 Rajit Manohar
 *  All Rights Reserved
 *
 **************************************************************************
 */
#include <stdio.h>
#include <ctype.h>
#include <act/act.h>
#include "act_parse.h"
#include "act_walk_X.h"
#include "config.h"
#include "array.h"

#ifdef DEBUG_PERFORMANCE
#include "mytime.h"
#endif

struct command_line_defs {
  char *varname;
  int value;
};

int Act::max_recurse_depth;

L_A_DECL (struct command_line_defs, vars);

#define isid(x) (((x) == '_') || isalpha(x))

/*------------------------------------------------------------------------
 *
 *  _valid_id_string --
 *
 *   Returns 1 if it is a valid id string, 0 otherwise
 *
 *------------------------------------------------------------------------
 */
static int 
_valid_id_string (char *s)
{
  if (!*s) return 0;
  if (!isid (*s)) return 0;
  s++;
  while (*s) {
    if (!isid (*s) && !isdigit (*s)) return 0;
    s++;
  }
  return 1;
}


void Act::Init (int *iargc, char ***iargv)
{
  static int initialize = 0;
  int argc = *iargc;
  char **argv = *iargv;
  int i, j;

  if (initialize) return;
  initialize = 1;
  ActNamespace::Init();
  Type::Init();

  config_std_path ("act");
  config_set_default_int ("act.max_recurse_depth", 1000);
  config_read ("global.conf");
  Act::max_recurse_depth = config_get_int ("act.max_recurse_depth");

  A_INIT (vars);

  for (i=1; i < argc; i++) {
    if (strncmp (argv[i], "-D", 2) == 0) {
      char *tmp, *s;

      tmp = Strdup (argv[i]+2);
      /* tmp should look like var=true or var=false */
      s = tmp;
      while (*s && *s != '=') {
	s++;
      }

      if (*s != '=') {
	fatal_error ("-D option must be of the form -Dvar=true or -Dvar=false");
      }
      *s = '\0';
      s++;
      if (strcmp (s, "true") != 0 &&
	  strcmp (s, "false") != 0) {
	fatal_error ("-D option must be of the form -Dvar=true or -Dvar=false");
      }

      if (!_valid_id_string (tmp)) {
	fatal_error ("-D option: variable `%s' is not a valid identifier",
		     tmp);
      }

      A_NEW (vars, struct command_line_defs);
      A_NEXT (vars).varname = tmp;
      if (strcmp (s, "true") == 0) {
	A_NEXT (vars).value = 1;
      }
      else {
	A_NEXT (vars).value = 0;
      }
      A_INC (vars);
    }
    else if (strncmp (argv[i], "-T", 2) == 0) {
      config_stdtech_path (argv[i]+2);
    }
    else {
      break;
    }
  }
  /* arguments from position 1 ... 1 + (i-1) have been wacked */
  *iargc = (argc - i + 1);
  for (j=1; j < *iargc; j++) {
    argv[j] = argv[j+(i-1)];
  }
  argv[j] = NULL;
  return;
}


Act::Act (const char *s)
{
  act_Token *a;
  ActTree tr;
  int argc = 1;
  char **argv;

  MALLOC (argv, char *, 2);
  argv[0] = Strdup ("-internal-");
  argv[1] = NULL;
  Act::Init (&argc, &argv);
  FREE (argv[0]);
  FREE (argv);

  aux = hash_new (2);

  gns = ActNamespace::global;
  tf = new TypeFactory();
  
  if (!s) {
    return;
  }

  tr.curns = ActNamespace::global;
  tr.global = tr.curns;
  tr.os = new ActOpen ();
  tr.tf = tf;
  tr.scope = tr.curns->CurScope ();

  tr.u = NULL;

  tr.param_mode = 0;

  tr.u_p = NULL;
  tr.u_d = NULL;
  tr.u_c = NULL;
  tr.u_f = NULL;
  tr.t = NULL;

  tr.t_inst = NULL;

  tr.i_t = NULL;
  tr.i_id = NULL;
  tr.a_id = NULL;

  tr.supply.vdd = NULL;
  tr.supply.gnd = NULL;
  tr.supply.psc = NULL;
  tr.supply.nsc = NULL;
  
  tr.strict_checking = 0;

  tr.in_tree = 0;
  tr.in_subckt = 0;

  tr.attr_num = config_get_table_size ("act.instance_attr");
  tr.attr_table = config_get_table_string ("act.instance_attr");

#ifdef DEBUG_PERFORMANCE
  realtime_msec ();
#endif

  a = act_parse (s);

#ifdef DEBUG_PERFORMANCE
  printf ("Parser time: %g\n", (realtime_msec()/1000.0));
#endif

  ActBody *bhead, *prev;
  InstType *it;
  bhead = NULL;
  prev = NULL;
  for (int i=0; i < A_LEN (vars); i++) {
    ActBody *b;
    ActBody *bc;
    Expr *e;
    /* add variables to the global namespace
       Note: copied from instance[] rule in defs.m4
     */
    it = tr.tf->NewPBool();
    if (tr.global->CurScope()->Lookup (vars[i].varname)) {
      fatal_error ("Name `%s' defined through -D is a duplicate!", vars[i].varname);
    }
    tr.global->CurScope()->Add (vars[i].varname, it);
    b = new ActBody_Inst (it, vars[i].varname);
    NEW (e, Expr);
    e->type = (vars[i].value ? E_TRUE : E_FALSE);

    Expr *tmp = TypeFactory::NewExpr (e);
    FREE (e);
    e = tmp;

    bc = new ActBody_Conn (new ActId (vars[i].varname),
			   new AExpr (e));
    b->Append (bc);
    if (!bhead) {
      bhead = b;
      prev = bhead->Tail();
    }
    else {
      prev->Append (b);
      prev = prev->Tail ();
    }
  }
  if (bhead) {
    tr.global->AppendBody (bhead);
  }

  act_walk_X (&tr, a);
  
  act_parse_free (a);

#ifdef DEBUG_PERFORMANCE
  printf ("Walk and free time: %g\n", (realtime_msec()/1000.0));
#endif

  if (config_exists ("act.mangle_chars")) {
    mangle (config_get_string ("act.mangle_chars"));
  }
  ActNamespace::act = this;
}

void Act::Merge (const char *s)
{
  act_Token *a;
  ActTree tr;

  if (!s) {
    return;
  }

  tr.curns = ActNamespace::global;
  tr.global = tr.curns;
  tr.os = new ActOpen ();
  tr.tf = tf;
  tr.scope = tr.curns->CurScope ();

  tr.u = NULL;

  tr.param_mode = 0;

  tr.u_p = NULL;
  tr.u_d = NULL;
  tr.u_c = NULL;
  tr.u_f = NULL;
  tr.t = NULL;

  tr.t_inst = NULL;

  tr.i_t = NULL;
  tr.i_id = NULL;
  tr.a_id = NULL;

  tr.supply.vdd = NULL;
  tr.supply.gnd = NULL;
  tr.supply.psc = NULL;
  tr.supply.nsc = NULL;
  
  tr.strict_checking = 0;

  tr.in_tree = 0;
  tr.in_subckt = 0;

  tr.attr_num = config_get_table_size ("act.instance_attr");
  tr.attr_table = config_get_table_string ("act.instance_attr");

#ifdef DEBUG_PERFORMANCE
  realtime_msec ();
#endif

  a = act_parse (s);

#ifdef DEBUG_PERFORMANCE
  printf ("Parser time: %g\n", (realtime_msec()/1000.0));
#endif

  act_walk_X (&tr, a);
  
  act_parse_free (a);

#ifdef DEBUG_PERFORMANCE
  printf ("Walk and free time: %g\n", (realtime_msec()/1000.0));
#endif
}


void Act::Expand ()
{
  Assert (gns, "Expand() called without an object?");
  /* expand each namespace! */
  gns->Expand ();
}



Process *Act::findProcess (const char *s)
{
  if (!s) return NULL;
  
  int i;
  char *tmp = Strdup (s);
  char *f = tmp;
  ActNamespace *ns = gns;

  i = 0;
  while (tmp[i]) {
    if (tmp[i] == ':') {
      if (tmp[i+1] != ':') {
	return NULL;
      }
      else {
	tmp[i] = '\0';
	if (i == 0) {
	  ns = gns;
	}
	else {
	  ns = ns->findNS (tmp);
	}
	if (!ns) {
	  return NULL;
	}
	tmp = tmp + i + 2;
	i = 0;
      }
    }
    else {
      i++;
    }
  }

  Process *p = findProcess (ns, tmp);
  FREE (f);
  return p;
}

Process *Act::findProcess (ActNamespace *n, const char *s)
{
  UserDef *u;

  if (!s || !n) return NULL;

  u = n->findType (s);
  if (!u) return NULL;
  return dynamic_cast<Process *>(u);
}


ActNamespace *Act::findNamespace (const char *s)
{
  if (!s) return NULL;
  
  UserDef *u;
  int i;
  char *tmp = Strdup (s);
  char *f = tmp;
  ActNamespace *ns = gns;

  i = 0;
  while (tmp[i]) {
    if (tmp[i] == ':') {
      if (tmp[i+1] != ':') {
	return NULL;
      }
      else {
	tmp[i] = '\0';
	if (i == 0) {
	  ns = gns;
	}
	else {
	  ns = ns->findNS (tmp);
	}
	if (!ns) {
	  return NULL;
	}
	tmp = tmp + i + 2;
	i = 0;
      }
    }
    else {
      i++;
    }
  }

  ns = ns->findNS (tmp);
  
  FREE (f);

  return ns;
}


ActNamespace *Act::findNamespace (ActNamespace *ns, const char *s)
{
  if (!ns || !s) return NULL;
  return ns->findNS (s);
}
		 

void Act::aux_add (const char *phase, void *data)
{
  hash_bucket_t *b;

  b = hash_lookup (aux, phase);
  if (b) {
    if (b->v) {
      warning ("Act::aux_append is replacing old data for `%s'", phase);
    }
    b->v = data;
  }
  else {
    b = hash_add (aux, phase);
    b->v = data;
  }
}

void *Act::aux_find (const char *phase)
{
  hash_bucket_t *b;

  b = hash_lookup (aux, phase);
  if (b) {
    return b->v;
  }
  else {
    return NULL;
  }
}


void Act::Print (FILE *fp)
{
  gns->Print (fp);
}
