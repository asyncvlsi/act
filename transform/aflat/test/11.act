import "env.act";

env::source_e1of2<6,{0,1,1,0,0,1}> src;
env::sink_e1of2 snk(src.r);
