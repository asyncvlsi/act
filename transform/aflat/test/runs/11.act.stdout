~"Reset"&("snk.l.f"|"snk.l.t")->"snk.l.e"-
"Reset"|~"snk.l.f"&~"snk.l.t"->"snk.l.e"+
= "snk.l.d[0]" "snk.l.f"
= "snk.l.d[1]" "snk.l.t"
= "snk.l.f" "src.r.f"
= "snk.l.d[0]" "src.r.d[0]"
= "snk.l.d[1]" "src.r.d[1]"
= "snk.l.e" "src.r.e"
= "snk.l.t" "src.r.t"
"src.src.s.d[0]"->"src.src.r.d[0]"+
"src.src.s.d[1]"->"src.src.r.d[1]"+
"src.src.s.d[2]"->"src.src.r.d[1]"+
"src.src.s.d[3]"->"src.src.r.d[0]"+
"src.src.s.d[4]"->"src.src.r.d[0]"+
"src.src.s.d[5]"->"src.src.r.d[1]"+
~"src.src.s.d[0]"&~"src.src.s.d[1]"&~"src.src.s.d[2]"&~"src.src.s.d[3]"&~"src.src.s.d[4]"&~"src.src.s.d[5]"->"src.src.r.d[0]"-
~"src.src.s.d[0]"&~"src.src.s.d[1]"&~"src.src.s.d[2]"&~"src.src.s.d[3]"&~"src.src.s.d[4]"&~"src.src.s.d[5]"->"src.src.r.d[1]"-
"src.src.r.d[0]"->"src.src.rr.d[0]"+
"src.src.r.d[1]"->"src.src.rr.d[1]"+
~"src.src.r.d[0]"->"src.src.rr.d[0]"-
~"src.src.r.d[1]"->"src.src.rr.d[1]"-
= "src.src.rr.e" "src.src.sc[5].se"
= "src.src.rr.e" "src.src.sc[4].se"
= "src.src.rr.e" "src.src.sc[3].se"
= "src.src.rr.e" "src.src.sc[2].se"
= "src.src.rr.e" "src.src.sc[1].se"
= "src.src.rr.e" "src.src.sc[0].se"
= "src.src.rr.e" "src.src.s.e"
~"Reset"&~"src.src.sc[0].x"&~"src.src.sc[0].y"&"src.src.sc[0].l"->"src.src.sc[0].le"-
~"Reset"&~"src.src.sc[0].y"&~"src.src.sc[0].le"->"src.src.sc[0].x"+
"Reset"|"src.src.sc[0].x"&~"src.src.sc[0].y"&~"src.src.sc[0].l"->"src.src.sc[0].le"+
~"Reset"&"src.src.sc[0].x"&"src.src.sc[0].le"->"src.src.sc[0].y"+
~"Reset"&"src.src.sc[0].x"&"src.src.sc[0].y"&"src.src.sc[0].se"&"src.src.sc[0].re"->"src.src.sc[0].s"+
"Reset"|"src.src.sc[0].y"&"src.src.sc[0].s"->"src.src.sc[0].x"-
"Reset"|~"src.src.sc[0].x"&"src.src.sc[0].y"&~"src.src.sc[0].se"&~"src.src.sc[0].re"->"src.src.sc[0].s"-
"Reset"|~"src.src.sc[0].x"&~"src.src.sc[0].s"->"src.src.sc[0].y"-
= "src.src.sc[0].s" "src.src.sc[0].r"
~"Reset"&~"src.src.sc[1].x"&~"src.src.sc[1].y"&"src.src.sc[1].l"->"src.src.sc[1].le"-
~"Reset"&~"src.src.sc[1].y"&~"src.src.sc[1].le"->"src.src.sc[1].x"+
"Reset"|"src.src.sc[1].x"&~"src.src.sc[1].y"&~"src.src.sc[1].l"->"src.src.sc[1].le"+
~"Reset"&"src.src.sc[1].x"&"src.src.sc[1].le"->"src.src.sc[1].y"+
~"Reset"&"src.src.sc[1].x"&"src.src.sc[1].y"&"src.src.sc[1].se"&"src.src.sc[1].re"->"src.src.sc[1].s"+
"Reset"|"src.src.sc[1].y"&"src.src.sc[1].s"->"src.src.sc[1].x"-
"Reset"|~"src.src.sc[1].x"&"src.src.sc[1].y"&~"src.src.sc[1].se"&~"src.src.sc[1].re"->"src.src.sc[1].s"-
"Reset"|~"src.src.sc[1].x"&~"src.src.sc[1].s"->"src.src.sc[1].y"-
= "src.src.sc[1].s" "src.src.sc[1].r"
~"Reset"&~"src.src.sc[2].x"&~"src.src.sc[2].y"&"src.src.sc[2].l"->"src.src.sc[2].le"-
~"Reset"&~"src.src.sc[2].y"&~"src.src.sc[2].le"->"src.src.sc[2].x"+
"Reset"|"src.src.sc[2].x"&~"src.src.sc[2].y"&~"src.src.sc[2].l"->"src.src.sc[2].le"+
~"Reset"&"src.src.sc[2].x"&"src.src.sc[2].le"->"src.src.sc[2].y"+
~"Reset"&"src.src.sc[2].x"&"src.src.sc[2].y"&"src.src.sc[2].se"&"src.src.sc[2].re"->"src.src.sc[2].s"+
"Reset"|"src.src.sc[2].y"&"src.src.sc[2].s"->"src.src.sc[2].x"-
"Reset"|~"src.src.sc[2].x"&"src.src.sc[2].y"&~"src.src.sc[2].se"&~"src.src.sc[2].re"->"src.src.sc[2].s"-
"Reset"|~"src.src.sc[2].x"&~"src.src.sc[2].s"->"src.src.sc[2].y"-
= "src.src.sc[2].s" "src.src.sc[2].r"
~"Reset"&~"src.src.sc[3].x"&~"src.src.sc[3].y"&"src.src.sc[3].l"->"src.src.sc[3].le"-
~"Reset"&~"src.src.sc[3].y"&~"src.src.sc[3].le"->"src.src.sc[3].x"+
"Reset"|"src.src.sc[3].x"&~"src.src.sc[3].y"&~"src.src.sc[3].l"->"src.src.sc[3].le"+
~"Reset"&"src.src.sc[3].x"&"src.src.sc[3].le"->"src.src.sc[3].y"+
~"Reset"&"src.src.sc[3].x"&"src.src.sc[3].y"&"src.src.sc[3].se"&"src.src.sc[3].re"->"src.src.sc[3].s"+
"Reset"|"src.src.sc[3].y"&"src.src.sc[3].s"->"src.src.sc[3].x"-
"Reset"|~"src.src.sc[3].x"&"src.src.sc[3].y"&~"src.src.sc[3].se"&~"src.src.sc[3].re"->"src.src.sc[3].s"-
"Reset"|~"src.src.sc[3].x"&~"src.src.sc[3].s"->"src.src.sc[3].y"-
= "src.src.sc[3].s" "src.src.sc[3].r"
~"Reset"&~"src.src.sc[4].x"&~"src.src.sc[4].y"&"src.src.sc[4].l"->"src.src.sc[4].le"-
~"Reset"&~"src.src.sc[4].y"&~"src.src.sc[4].le"->"src.src.sc[4].x"+
"Reset"|"src.src.sc[4].x"&~"src.src.sc[4].y"&~"src.src.sc[4].l"->"src.src.sc[4].le"+
~"Reset"&"src.src.sc[4].x"&"src.src.sc[4].le"->"src.src.sc[4].y"+
~"Reset"&"src.src.sc[4].x"&"src.src.sc[4].y"&"src.src.sc[4].se"&"src.src.sc[4].re"->"src.src.sc[4].s"+
"Reset"|"src.src.sc[4].y"&"src.src.sc[4].s"->"src.src.sc[4].x"-
"Reset"|~"src.src.sc[4].x"&"src.src.sc[4].y"&~"src.src.sc[4].se"&~"src.src.sc[4].re"->"src.src.sc[4].s"-
"Reset"|~"src.src.sc[4].x"&~"src.src.sc[4].s"->"src.src.sc[4].y"-
= "src.src.sc[4].s" "src.src.sc[4].r"
~"Reset"&~"src.src.sc[5].x"&~"src.src.sc[5].y"&"src.src.sc[5].l"->"src.src.sc[5].le"-
~"Reset"&~"src.src.sc[5].y"&~"src.src.sc[5].le"->"src.src.sc[5].x"+
"Reset"|"src.src.sc[5].x"&~"src.src.sc[5].y"&~"src.src.sc[5].l"->"src.src.sc[5].le"+
~"Reset"&"src.src.sc[5].x"&"src.src.sc[5].le"->"src.src.sc[5].y"+
~"Reset"&"src.src.sc[5].x"&"src.src.sc[5].y"&"src.src.sc[5].se"&"src.src.sc[5].re"->"src.src.sc[5].s"+
"Reset"|"src.src.sc[5].y"&"src.src.sc[5].s"->"src.src.sc[5].x"-
"Reset"|~"src.src.sc[5].x"&"src.src.sc[5].y"&~"src.src.sc[5].se"&~"src.src.sc[5].re"->"src.src.sc[5].s"-
"Reset"|~"src.src.sc[5].x"&~"src.src.sc[5].s"->"src.src.sc[5].y"-
= "src.src.sc[5].s" "src.src.sc[5].r"
= "src.src.l[0]" "src.src.startc.r"
= "src.src.l[0]" "src.src.sc[0].l"
= "src.src.l[1]" "src.src.sc[1].l"
= "src.src.l[1]" "src.src.sc[0].s"
= "src.src.l[1]" "src.src.s.d[0]"
= "src.src.l[2]" "src.src.sc[2].l"
= "src.src.l[2]" "src.src.sc[1].s"
= "src.src.l[2]" "src.src.s.d[1]"
= "src.src.l[3]" "src.src.sc[3].l"
= "src.src.l[3]" "src.src.sc[2].s"
= "src.src.l[3]" "src.src.s.d[2]"
= "src.src.l[4]" "src.src.sc[4].l"
= "src.src.l[4]" "src.src.sc[3].s"
= "src.src.l[4]" "src.src.s.d[3]"
= "src.src.l[5]" "src.src.sc[5].l"
= "src.src.l[5]" "src.src.sc[4].s"
= "src.src.l[5]" "src.src.s.d[4]"
= "src.src.l[6]" "src.src.startc.l"
= "src.src.l[6]" "src.src.sc[5].s"
= "src.src.l[6]" "src.src.s.d[5]"
~"Reset"&~"src.src.startc.x"&~"src.src.startc.y"&"src.src.startc.re"->"src.src.startc.r"+
~"Reset"&~"src.src.startc.y"&"src.src.startc.r"->"src.src.startc.x"+
"Reset"|"src.src.startc.x"&~"src.src.startc.y"&~"src.src.startc.re"->"src.src.startc.r"-
~"Reset"&"src.src.startc.x"&~"src.src.startc.r"->"src.src.startc.y"+
~"Reset"&"src.src.startc.x"&"src.src.startc.y"&"src.src.startc.l"->"src.src.startc.le"-
"Reset"|"src.src.startc.y"&~"src.src.startc.le"->"src.src.startc.x"-
"Reset"|~"src.src.startc.x"&"src.src.startc.y"&~"src.src.startc.l"->"src.src.startc.le"+
"Reset"|~"src.src.startc.x"&"src.src.startc.le"->"src.src.startc.y"-
= "src.src.le[0]" "src.src.startc.re"
= "src.src.le[0]" "src.src.sc[0].le"
= "src.src.le[1]" "src.src.sc[1].le"
= "src.src.le[1]" "src.src.sc[0].re"
= "src.src.le[2]" "src.src.sc[2].le"
= "src.src.le[2]" "src.src.sc[1].re"
= "src.src.le[3]" "src.src.sc[3].le"
= "src.src.le[3]" "src.src.sc[2].re"
= "src.src.le[4]" "src.src.sc[4].le"
= "src.src.le[4]" "src.src.sc[3].re"
= "src.src.le[5]" "src.src.sc[5].le"
= "src.src.le[5]" "src.src.sc[4].re"
= "src.src.le[6]" "src.src.startc.le"
= "src.src.le[6]" "src.src.sc[5].re"
= "src.s.d[0]" "src.src.rr.d[0]"
= "src.s.d[1]" "src.src.rr.d[1]"
= "src.s.e" "src.src.rr.e"
= "src.r.d[0]" "src.r.f"
= "src.r.d[1]" "src.r.t"
= "src.r.d[0]" "src.s.d[0]"
= "src.r.d[1]" "src.s.d[1]"
= "src.r.e" "src.s.e"
